File: cpp/src/io.cpp
#include "io.hpp"

#include <algorithm>
#include <fstream>
#include <sstream>
#include <stdexcept>

Instance load_instance(const std::filesystem::path& path) {
    std::ifstream input(path);
    if (!input) {
        throw std::runtime_error("could not open instance file: " + path.string());
    }

    std::string header;
    if (!std::getline(input, header)) {
        throw std::runtime_error("missing instance header: " + path.string());
    }

    std::replace(header.begin(), header.end(), ',', ' ');
    std::istringstream header_stream(header);

    Instance instance;
    if (!(header_stream >> instance.n >> instance.a >> instance.b)) {
        throw std::runtime_error("invalid instance header: " + path.string());
    }

    instance.matrix.assign(instance.n * instance.n, 0.0);
    for (std::size_t i = 0; i < instance.n; ++i) {
        std::string line;
        if (!std::getline(input, line)) {
            throw std::runtime_error("missing matrix row in: " + path.string());
        }

        std::istringstream row_stream(line);
        for (std::size_t j = 0; j < instance.n; ++j) {
            if (!(row_stream >> instance.matrix[i * instance.n + j])) {
                throw std::runtime_error("invalid matrix row in: " + path.string());
            }
        }
    }

    return instance;
}

std::vector<std::filesystem::path> csv_instance_files(const std::filesystem::path& directory) {
    std::vector<std::filesystem::path> files;

    for (const auto& entry : std::filesystem::directory_iterator(directory)) {
        if (entry.is_regular_file() && entry.path().extension() == ".csv") {
            files.push_back(entry.path());
        }
    }

    std::sort(files.begin(), files.end());
    return files;
}

std::string clean_filename(const std::string& path) {
    auto name = std::filesystem::path(path).filename().string();
    return name.empty() ? path : name;
}

std::filesystem::path repo_root() {
#ifdef PROJECT_ROOT
    return std::filesystem::path(PROJECT_ROOT);
#else
    auto current = std::filesystem::current_path();
    while (!current.empty()) {
        if (std::filesystem::exists(current / "instances") &&
            std::filesystem::exists(current / "results")) {
            return current;
        }
        current = current.parent_path();
    }
    return std::filesystem::current_path();
#endif
}

File: rust/src/io.rs
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

#[derive(Clone, Debug)]
pub struct Instance {
    pub matrix: Vec<f64>,
    pub n: usize,
    pub a: usize,
    pub b: usize,
}

pub fn load_instance(path: impl AsRef<Path>) -> io::Result<Instance> {
    let content = fs::read_to_string(path)?;
    let mut lines = content.lines();

    let header = lines
        .next()
        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, "missing instance header"))?;
    let mut header_parts = header.split(',').map(str::trim);

    let n = parse_usize(header_parts.next(), "n")?;
    let a = parse_usize(header_parts.next(), "a")?;
    let b = parse_usize(header_parts.next(), "b")?;

    let mut matrix = vec![0.0; n * n];
    for i in 0..n {
        let line = lines.next().ok_or_else(|| {
            io::Error::new(
                io::ErrorKind::InvalidData,
                format!("missing matrix row {i}"),
            )
        })?;
        for (j, value) in line.split_whitespace().enumerate().take(n) {
            matrix[i * n + j] = value.parse::<f64>().map_err(|err| {
                io::Error::new(
                    io::ErrorKind::InvalidData,
                    format!("invalid matrix value at row {i}, column {j}: {err}"),
                )
            })?;
        }
    }

    Ok(Instance { matrix, n, a, b })
}

pub fn csv_instance_files(dir: impl AsRef<Path>) -> io::Result<Vec<PathBuf>> {
    let mut files = Vec::new();
    for entry in fs::read_dir(dir)? {
        let path = entry?.path();
        if path.extension().is_some_and(|ext| ext == "csv") {
            files.push(path);
        }
    }
    files.sort();
    Ok(files)
}

pub fn clean_filename(path: impl AsRef<str>) -> String {
    Path::new(path.as_ref())
        .file_name()
        .and_then(|name| name.to_str())
        .unwrap_or(path.as_ref())
        .to_string()
}

fn parse_usize(value: Option<&str>, name: &str) -> io::Result<usize> {
    value
        .ok_or_else(|| io::Error::new(io::ErrorKind::InvalidData, format!("missing {name}")))?
        .parse::<usize>()
        .map_err(|err| {
            io::Error::new(
                io::ErrorKind::InvalidData,
                format!("invalid {name} in header: {err}"),
            )
        })
}
